backend/src/main/java/ai/verta/modeldb/lineage/LineageDAORdbImpl.java

```diff
@@ -30,36 +49,241 @@
+        lineageElementEntity = session.get(LineageElementEntity.class, addLineage.getId());
+      } else {
+        id = null;
+        lineageElementEntity = null;
+      }
+      if (lineageElementEntity == null) {
+        lineageElementEntity = new LineageElementEntity(id);
+        session.save(lineageElementEntity);
+        id = lineageElementEntity.getId();
+      }
       validate(addLineage.getInputList(), addLineage.getOutputList());
-      validateExistence(
-          addLineage.getInputList(), addLineage.getOutputList(), isExistsPredicate, session);
+      List<LineageEntry> lineageEntries = new LinkedList<>(addLineage.getInputList());
+      lineageEntries.addAll(addLineage.getOutputList());
+      existsCheckConsumer.test(session, lineageEntries);
       for (LineageEntry input : addLineage.getInputList()) {
-        for (LineageEntry output : addLineage.getOutputList()) {
-          addLineage(session, input, output);
-        }
+        addLineage(session, input, id, CONNECTION_TYPE_INPUT, blobHashInCommitFunction);
+      }
+      for (LineageEntry output : addLineage.getOutputList()) {
+        addLineage(session, output, id, CONNECTION_TYPE_OUTPUT, blobHashInCommitFunction);
       }
       session.getTransaction().commit();
+      return AddLineage.Response.newBuilder().setId(lineageElementEntity.getId()).build();
     }
-    return AddLineage.Response.newBuilder().setStatus(true).build();
   }
 
   @Override
-  public DeleteLineage.Response deleteLineage(DeleteLineage deleteLineage) throws ModelDBException {
+  public DeleteLineage.Response deleteLineage(
+      DeleteLineage deleteLineage, BlobHashInCommitFunction blobHashInCommitFunction)
+      throws ModelDBException {
+    boolean status;
     try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
       session.beginTransaction();
       validate(deleteLineage.getInputList(), deleteLineage.getOutputList());
+      long id = deleteLineage.getId();
+      if (id == 0) {
+        throw new ModelDBException("Id not specified", Code.INVALID_ARGUMENT);
+      }
+      List connectionEntitiesById = getConnectionEntitiesById(session, id);
+      Map.Entry<Map<LineageElement, ConnectionEntity>, Map<LineageElement, ConnectionEntity>>
+          inputOutputs = getInputOutputs(session, connectionEntitiesById);
+      Map<LineageElement, ConnectionEntity> inputInDatabase = inputOutputs.getKey();
+      Map<LineageElement, ConnectionEntity> outputInDatabase = inputOutputs.getValue();
       for (LineageEntry input : deleteLineage.getInputList()) {
-        for (LineageEntry output : deleteLineage.getOutputList()) {
-          deleteLineage(session, input, output);
+        deleteLineage(session, input, inputInDatabase, blobHashInCommitFunction);
+      }
+      for (LineageEntry output : deleteLineage.getOutputList()) {
+        deleteLineage(session, output, outputInDatabase, blobHashInCommitFunction);
+      }
+      status = false;
```

Comment:
Status should never be false. I don't even know status of what this is, but I assume it's there for historical reasons.

Reply:
fixed